<?xml version="1.0"?>
<launch>
    <arg name="world" default="empty"/>
    <arg name="scale" default="1.0"/>
    <arg name="gui" default="false"/>
    <arg name="x" default="0.0"/>
    <arg name="y" default="0.0"/>
    <arg name="z" default="0.1"/>
    <arg name="nodelet_manager_name" default="nodelet_manager_control"/>
    <arg name="start_nodelet_manager" default="true"/>

  <arg name="robot" default="/"/>
  <arg name="odom" default="/mavros/local_position/odom"/>
  <arg name="so3_cmd" default="so3_cmd"/>
  <arg name="num_props" default="4"/>
  <arg name="kf" default="8.54858e-06"/>
  <arg name="lin_cof_a" default="0.001798354"/>
  <arg name="lin_int_b" default="-0.618518519"/>

    <!-- Launch rviz -->
    <node pkg="rviz" type="rviz" name="rviz"
          args="-d $(find erl_quadrotor_control)/rviz/quadrotor.rviz"/>


    <node pkg="nodelet"
      type="nodelet"
      name="$(arg nodelet_manager_name)"
      args="manager"
      output="screen"
      if="$(arg start_nodelet_manager)"/>

    <!-- Launch controller -->
    <node pkg="nodelet"
      type="nodelet"
      args="standalone kr_mav_controllers/SO3ControlNodelet"
      name="so3_control"
      required="true"
      output="screen">
      <rosparam file="$(find erl_quadrotor_control)/config/gains.yaml"/>
      <param name="mass" value="1.5"/>
      <remap from="~odom" to="/mavros/local_position/odom"/>
      <remap from="~position_cmd" to="position_cmd"/>
      <remap from="~so3_cmd" to="so3_cmd"/>
      <remap from="~motors" to="motors"/>
    </node>

    <node pkg="nodelet"
      type="nodelet"
      args="standalone kr_mavros_interface/SO3CmdToMavros"
      name="so3cmd_to_mavros"
      required="true"
      clear_params="true"
      output="screen">
      <param name="num_props" value="$(arg num_props)"/>
      <param name="kf" value="$(arg kf)"/>
      <param name="lin_cof_a" value="$(arg lin_cof_a)"/>
      <param name="lin_int_b" value="$(arg lin_int_b)"/>
      <remap from="~odom" to="$(arg odom)"/>
      <remap from="~so3_cmd" to="$(arg so3_cmd)"/>
      <remap from="~imu" to="mavros/imu/data" />
      <remap from="~attitude_raw" to="mavros/setpoint_raw/attitude" />
      <remap from="~odom_pose" to="mavros/vision_pose/pose" />
    </node>


    <!-- Launch PX4 simulator -->
    <include file="$(find px4)/launch/mavros_posix_sitl.launch">
    </include>

    <!-- Launch offboard node -->
    <node name="offb_geoctl" pkg="erl_quadrotor_control" type="offb_geoctl_tracking_node" output="screen"/>
</launch>
